# Data Cleaning

## Orders DataFrame

In [16]:
# 1. Import libraries
import pandas as pd
import numpy as np
import os

In [17]:
# Path to the original data
path= r"/Users/mentaykoshzhanova/Desktop/Instacart Basket Analysis"
path

'/Users/mentaykoshzhanova/Desktop/Instacart Basket Analysis'

In [18]:
# Import dataframes
df_prods= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

In [19]:
# Running describe for products df 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [22]:
# 2. Running describe for orders df 
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  order_time  \
0           0   2539329        1             1                  2           8   
1           1   2398795        1             2                  3           7   
2           2    473747        1             3                  3          12   
3           3   2254736        1             4                  4           7   
4           4    431534        1             5                  4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [24]:
df_ords.drop(df_ords.columns[[0]], axis=1)

order_id  user_id  order_number  order_day_of_week  order_time  \
0         2539329        1             1                  2           8   
1         2398795        1             2                  3           7   
2          473747        1             3                  3          12   
3         2254736        1             4                  4           7   
4          431534        1             5                  4          15   
...           ...      ...           ...                ...         ...   
3421078   2266710   206209            10                  5          18   
3421079   1854736   206209            11                  4          10   
3421080    626363   206209            12                  1          12   
3421081   2977660   206209            13                  1          12   
3421082    272231   206209            14                  6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [25]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week    order_time  days_since_prior_order  
count       3.421083e+06  3.421083e+06            3.214874e+06  
mean        2.776219e+00  1.345202e+01            1.111484e+01  
std         2.046829e+00  4.226088e+00            9.206737e+00  
min         0.000000e+00  0.000000e+00            0.000000e+00  
25%         1.000000e+00  1.000000e+01            4.000000e+00  
50%         3.000000e+00  1.300000e+01            7.000000e+00  
75%         5.000000e+00  1.600000e+01            1.500000e+01  
max         6.000000e+00  2.300000e+01            3.000000e+01

### 2. Interpreting describe function
I would investigate further order_number column, because min and max are the same for this column showing 1.0

In [26]:
# 3. Checking orders df for mixed-data type
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird]>0):
        print(col)

In [27]:
# 4. No to mixed-data type in orders df

In [28]:
# 5. Checking orders df for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [29]:
df_missing=df_ords[df_ords['days_since_prior_order'].isnull()== True]
df_missing

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

## Missing values explanation:
There are missing values only in days_since_prior_order column, after creating a subset for days_since_prior_order column i see that missing values (NaN) are where order_number is 1 and user_id from 1 to 206209 (equal to missing columns), which means there are new customers when the app was launched (and there is no historical data for those customers).

### Missing values replacing explanation:
I decided not to replace missing values. Order number 1 is the first order for that particular customer and the history started with order 1. Even if it shows up as a missing data we know it is important to keep as it is.

In [30]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [31]:
# To see the average for days_since_prior_order
df_ords.mean(axis='index')

Unnamed: 0                1.710541e+06
order_id                  1.710542e+06
user_id                   1.029782e+05
order_number              1.715486e+01
order_day_of_week         2.776219e+00
order_time                1.345202e+01
days_since_prior_order    1.111484e+01
dtype: float64

In [32]:
# 7. Checking current # of rows before finding duplicates
df_ords.shape

(3421083, 7)

In [33]:
# Finding duplicates in orders df
df_dups=df_ords[df_ords.duplicated()]
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_time, days_since_prior_order]
Index: []

In [34]:
# No duplicates in orders df, but running drop duplicates to be sure and checking # of rows after dropping
df_ords_clean=df_ords.drop_duplicates()
df_ords_clean.shape

(3421083, 7)

### 8. Duplicates
There is no duplicates showed up after running the fucntion. 

In [35]:
# 9. Exporting orders as csv
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean.csv'))